File: internal/awg/vpnuri/vpnuri_test.go
// Copyright (c) 2025 LucX-UI Project.
// Licensed under the PolyForm Noncommercial License 1.0.0.
// SPDX-License-Identifier: PolyForm-Noncommercial-1.0.0

package vpnuri

import (
	"encoding/json"
	"strings"
	"testing"
)

const sampleConf = `[Interface]
PrivateKey = CKLAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEE=
Address = 10.200.0.2/32
DNS = 1.1.1.1, 1.0.0.1
MTU = 1320
Jc = 4

[Peer]
PublicKey = DGSYIcEKAUkA7HhzGSjxLZuV67BR3LeyU0BMLJzNVHQ=
AllowedIPs = 0.0.0.0/0, ::/0
Endpoint = 1.2.3.4:51820
`

func TestEncodeDecodeRoundTrip(t *testing.T) {
	uri, err := EncodeConf(sampleConf)
	if err != nil {
		t.Fatal(err)
	}
	if !strings.HasPrefix(uri, "vpn://") {
		t.Fatalf("prefix: %s", uri[:20])
	}
	if strings.Contains(uri, "=") {
		t.Fatalf("unexpected padding in %s", uri)
	}
	got, err := Decode(uri)
	if err != nil {
		t.Fatal(err)
	}
	conf, err := ConfFromPayload(got)
	if err != nil {
		t.Fatal(err)
	}
	if strings.TrimSpace(conf) != strings.TrimSpace(sampleConf) {
		t.Fatalf("round-trip mismatch:\n got %q\nwant %q", conf, sampleConf)
	}
}

func TestEncodeConf_OfficialContainer(t *testing.T) {
	uri, err := EncodeConf("# alice\n" + sampleConf)
	if err != nil {
		t.Fatal(err)
	}
	raw, err := Decode(uri)
	if err != nil {
		t.Fatal(err)
	}
	var env map[string]any
	if err := json.Unmarshal(raw, &env); err != nil {
		t.Fatalf("payload must be JSON: %v\n%s", err, raw)
	}
	if env["defaultContainer"] != containerName {
		t.Fatalf("defaultContainer = %v", env["defaultContainer"])
	}
	if env["hostName"] != "1.2.3.4" {
		t.Fatalf("hostName = %v", env["hostName"])
	}
	if env["dns1"] != "1.1.1.1" || env["dns2"] != "1.0.0.1" {
		t.Fatalf("dns = %v / %v", env["dns1"], env["dns2"])
	}
	if env["description"] != "alice" {
		t.Fatalf("description = %v", env["description"])
	}
	containers, _ := env["containers"].([]any)
	if len(containers) != 1 {
		t.Fatalf("containers len = %d", len(containers))
	}
	c0, _ := containers[0].(map[string]any)
	if c0["container"] != containerName {
		t.Fatalf("container = %v", c0["container"])
	}
	awg, _ := c0["awg"].(map[string]any)
	if awg == nil {
		t.Fatal("missing awg key (NekoBox+ requires it)")
	}
	lc, _ := awg["last_config"].(string)
	if lc == "" {
		t.Fatal("last_config must be a JSON string")
	}
	var inner map[string]any
	if err := json.Unmarshal([]byte(lc), &inner); err != nil {
		t.Fatalf("last_config not JSON: %v", err)
	}
	cfg, _ := inner["config"].(string)
	if !strings.Contains(cfg, "[Interface]") || !strings.Contains(cfg, "Endpoint = 1.2.3.4:51820") {
		t.Fatalf("inner config missing .conf:\n%s", cfg)
	}
	if inner["client_priv_key"] != "CKLAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEE=" {
		t.Fatalf("client_priv_key = %v", inner["client_priv_key"])
	}
	if inner["client_ip"] != "10.200.0.2/32" {
		t.Fatalf("client_ip = %v", inner["client_ip"])
	}
	if inner["server_pub_key"] != "DGSYIcEKAUkA7HhzGSjxLZuV67BR3LeyU0BMLJzNVHQ=" {
		t.Fatalf("server_pub_key = %v", inner["server_pub_key"])
	}
	if inner["Jc"] != "4" {
		t.Fatalf("Jc = %v", inner["Jc"])
	}
	if inner["mtu"] != "1320" {
		t.Fatalf("mtu = %v", inner["mtu"])
	}
	port, _ := inner["port"].(float64)
	if port != 51820 {
		t.Fatalf("port = %v", inner["port"])
	}
	ips, _ := inner["allowed_ips"].([]any)
	if len(ips) != 2 || ips[0] != "0.0.0.0/0" || ips[1] != "::/0" {
		t.Fatalf("allowed_ips = %v", inner["allowed_ips"])
	}
}

func TestEncodeConf_DomainEndpoint(t *testing.T) {
	conf := strings.Replace(sampleConf, "1.2.3.4:51820", "vpn.example.com:51820", 1)
	uri, err := EncodeConf(conf)
	if err != nil {
		t.Fatal(err)
	}
	raw, err := Decode(uri)
	if err != nil {
		t.Fatal(err)
	}
	var env map[string]any
	if err := json.Unmarshal(raw, &env); err != nil {
		t.Fatal(err)
	}
	if env["hostName"] != "vpn.example.com" {
		t.Fatalf("envelope hostName = %v", env["hostName"])
	}
	containers, _ := env["containers"].([]any)
	c0, _ := containers[0].(map[string]any)
	awg, _ := c0["awg"].(map[string]any)
	var inner map[string]any
	if err := json.Unmarshal([]byte(awg["last_config"].(string)), &inner); err != nil {
		t.Fatal(err)
	}
	if inner["hostName"] != "vpn.example.com" {
		t.Fatalf("last_config hostName = %v", inner["hostName"])
	}
}

func TestEncodeConf_OmitsEmptyAwgKeys(t *testing.T) {
	uri, err := EncodeConf(sampleConf)
	if err != nil {
		t.Fatal(err)
	}
	raw, err := Decode(uri)
	if err != nil {
		t.Fatal(err)
	}
	var env map[string]any
	if err := json.Unmarshal(raw, &env); err != nil {
		t.Fatal(err)
	}
	containers, _ := env["containers"].([]any)
	c0, _ := containers[0].(map[string]any)
	awg, _ := c0["awg"].(map[string]any)
	var inner map[string]any
	if err := json.Unmarshal([]byte(awg["last_config"].(string)), &inner); err != nil {
		t.Fatal(err)
	}
	for _, k := range []string{"S1", "H1", "I1", "HeaderProtectionKey", "psk_key", "persistent_keep_alive"} {
		if _, ok := inner[k]; ok {
			t.Fatalf("empty key %s must be omitted, got %v", k, inner[k])
		}
	}
}

func TestConfFromPayload_LegacyRawConf(t *testing.T) {
	got, err := ConfFromPayload([]byte(sampleConf))
	if err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(got, "[Peer]") {
		t.Fatalf("got %q", got)
	}
}

func TestEncodeConf_RejectsEmpty(t *testing.T) {
	if _, err := EncodeConf(""); err == nil {
		t.Fatal("expected error")
	}
	if _, err := EncodeConf("not a conf"); err == nil {
		t.Fatal("expected error for missing sections")
	}
}

func envelopeAwg(t *testing.T, conf string) map[string]any {
	t.Helper()
	uri, err := EncodeConf(conf)
	if err != nil {
		t.Fatal(err)
	}
	raw, err := Decode(uri)
	if err != nil {
		t.Fatal(err)
	}
	var env map[string]any
	if err := json.Unmarshal(raw, &env); err != nil {
		t.Fatal(err)
	}
	containers, _ := env["containers"].([]any)
	c0, _ := containers[0].(map[string]any)
	awg, _ := c0["awg"].(map[string]any)
	if awg == nil {
		t.Fatal("missing awg key")
	}
	return awg
}

func TestEncodeConf_ProtocolVersion(t *testing.T) {
	if pv, ok := envelopeAwg(t, sampleConf)["protocol_version"]; ok {
		t.Fatalf("v1 config must omit protocol_version, got %v", pv)
	}

	v2 := strings.Replace(sampleConf, "Jc = 4", "Jc = 4\nS3 = 10\nS4 = 5", 1)
	if got := envelopeAwg(t, v2)["protocol_version"]; got != "2" {
		t.Fatalf("S3/S4 config must carry protocol_version 2, got %v", got)
	}

	v3 := strings.Replace(sampleConf, "Jc = 4",
		"Jc = 4\nHeaderProtectionKey = MCPfRGcDGotJ6TcnIdDqsemj2cMIiGHnPUHM5ivXN18=", 1)
	if got := envelopeAwg(t, v3)["protocol_version"]; got != "3" {
		t.Fatalf("HPK config must carry protocol_version 3, got %v", got)
	}
}

File: frontend/src/test/vpnuri.test.ts
import { deflateSync } from 'node:zlib';
import { describe, expect, it } from 'vitest';

import {
  bytesFromBase64Url,
  inflateStored,
  isQCompress,
  vpnConfFromLink,
  vpnUriFromConf,
} from '@/lib/awg/vpnuri';

const sampleConf = `[Interface]
PrivateKey = CKLAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEE=
Address = 10.200.0.2/32
DNS = 1.1.1.1, 1.0.0.1
MTU = 1320
Jc = 4

# alice
[Peer]
PublicKey = DGSYIcEKAUkA7HhzGSjxLZuV67BR3LeyU0BMLJzNVHQ=
AllowedIPs = 0.0.0.0/0, ::/0
Endpoint = 1.2.3.4:51820
`;

function qCompress(data: Uint8Array): Uint8Array {
  const z = deflateSync(data);
  const out = new Uint8Array(4 + z.length);
  new DataView(out.buffer).setUint32(0, data.length);
  out.set(z, 4);
  return out;
}

function vpnUriFromBytes(bytes: Uint8Array): string {
  let bin = '';
  for (const b of bytes) bin += String.fromCharCode(b);
  return `vpn://${btoa(bin).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '')}`;
}

describe('vpnConfFromLink', () => {
  it('returns plain .conf from an upstream vpn:// payload', async () => {
    const link = vpnUriFromBytes(new TextEncoder().encode(sampleConf));
    await expect(vpnConfFromLink(link)).resolves.toContain('Endpoint = 1.2.3.4:51820');
  });

  it('inflates a LucX qCompress JSON envelope to the inner .conf', async () => {
    const inner = {
      config: sampleConf,
      client_priv_key: 'CKLAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEE=',
      Jc: '4',
    };
    const env = {
      defaultContainer: 'amnezia-awg',
      containers: [
        {
          container: 'amnezia-awg',
          awg: { last_config: JSON.stringify(inner), isThirdPartyConfig: true },
        },
      ],
    };
    const link = vpnUriFromBytes(qCompress(new TextEncoder().encode(JSON.stringify(env))));
    const conf = await vpnConfFromLink(link);
    expect(conf).toContain('[Interface]');
    expect(conf).toContain('Jc = 4');
    expect(conf).toContain('Endpoint = 1.2.3.4:51820');
    expect(conf).not.toMatch(/\uFFFD/);
  });

  it('returns empty for a non-vpn:// link', async () => {
    await expect(vpnConfFromLink('amneziawg://x')).resolves.toBe('');
  });

  // Cross-language fixture: this exact URI was produced by the Go encoder
  // internal/awg/vpnuri.EncodeConf(sampleConf). If the TS decoder stops
  // understanding the Go side (or vice versa), subscriptions and panel
  // exports diverge — AmneziaVPN would import one but not the other.
  const goEncodedUri =
    'vpn://AAADOHiclFJfb9o-FP0q0f29-pfmD1snSzxQilr-FLFRKm1cFLmJKV7DTWQ7dID47pMdRLWHSWv8Ep1z7tHxuT5CXpEViqQ2wJdHEG8vwI-gzONG6WImtN33K1qrF-BWN5JBKYzN8jMER4RRjsAROggMQZRl9SaLTNUGgS8RotCfq8jTnLufFUPISyXJZqr2w3EUJpETJldpgvDO11rtsle596r-eNL7p28w6LYmPqafXQ7JSr0WuVwh0kyrnbByLPdBN_iQLVKvKLQ0JugGf6ZGup3OHRr6w4LYXTyMEenhceGINIkQaZQH3aCDSIj0XyBKlUtEWs6k1D5a81yqvE12ezf_PswH497itXd9vznczX_-mvxonj5f33xLJ3K_iG4eJqPD9On-axeRem35w5kLdymeBb51pAEVdaXIuihhEqZhh3-KvyTtZjaVsVOxlb6sM-2JrW1aLD0r60pbhHaUIRipd1JndfN82dOHYsMJmPcEDt4TGFgtyDgsq3VlK-DQFDWc2PtjBQ5iS_KgxP_uxZ5WDAq5Fk1p-3-ROIHJtaqtqshxrnhgUJCJgcN5ay2QeMD1FwO7VOPBJEzDDpx-DwDRQfdV';

  it('decodes the Go-encoded envelope (cross-language parity)', async () => {
    const conf = await vpnConfFromLink(goEncodedUri);
    expect(conf).toContain('[Interface]');
    expect(conf).toContain('Jc = 4');
    expect(conf).toContain('Endpoint = 1.2.3.4:51820');
    expect(conf).not.toMatch(/\uFFFD/);
  });
});

describe('vpnUriFromConf (panel-side envelope encoder)', () => {
  it('round-trips its own output through the async decoder', async () => {
    const link = vpnUriFromConf(sampleConf);
    expect(link.startsWith('vpn://')).toBe(true);
    await expect(vpnConfFromLink(link)).resolves.toBe(sampleConf.trim());
  });

  it('rejects payloads without [Interface]/[Peer]', () => {
    expect(vpnUriFromConf('no sections here')).toBe('');
    expect(vpnUriFromConf('')).toBe('');
  });

  // Link labels inflate panel-minted envelopes synchronously (stored-block
  // zlib); Go/sub-service envelopes use real deflate and return null there.
  it('emits stored-block zlib that inflateStored reads synchronously', () => {
    const link = vpnUriFromConf(sampleConf);
    const bytes = bytesFromBase64Url(link.slice('vpn://'.length));
    expect(isQCompress(bytes)).toBe(true);
    const inflated = inflateStored(bytes);
    expect(inflated).not.toBeNull();
    const env = JSON.parse(new TextDecoder().decode(inflated as Uint8Array)) as {
      defaultContainer?: string;
      hostName?: string;
      description?: string;
      containers?: { awg?: { port?: string; protocol_version?: string } }[];
    };
    expect(env.defaultContainer).toBe('amnezia-awg');
    expect(env.hostName).toBe('1.2.3.4');
    expect(env.description).toBe('alice');
    expect(env.containers?.[0]?.awg?.port).toBe('51820');
    expect(env.containers?.[0]?.awg?.protocol_version).toBeUndefined();
  });

  it('marks AWG3 configs with protocol_version 3', () => {
    const v3 = sampleConf.replace('Jc = 4', 'Jc = 4\nHeaderProtectionKey = AAAA');
    const bytes = bytesFromBase64Url(vpnUriFromConf(v3).slice('vpn://'.length));
    const env = JSON.parse(new TextDecoder().decode(inflateStored(bytes) as Uint8Array)) as {
      containers?: { awg?: { protocol_version?: string } }[];
    };
    expect(env.containers?.[0]?.awg?.protocol_version).toBe('3');
  });
});
